### ISSUE TITLE
Conflicting out/ref argument set

### ISSUE THREAD

**Issue by dtchepak** [nsubstitute/NSubstitute.Analyzers#62]
It is possible to set argument values from `.Returns` and `.AndDoes`. At present the one set in `.Returns` wins (not sure we should always rely on this for future versions), but this can result in confusing results.

```csharp
        public interface ILookup { bool Lookup(string key, out int i); }

        [Fact]
        public void Test() {
            var sub = Substitute.For<ILookup>();

            sub.Lookup("blah", out Arg.Any<int>())
               .Returns(x => {
                   x[1] = 42;
                   return true;
                })
               .AndDoes(x => x[1] = 45);

            sub.Lookup("blah", out var value);

            Assert.Equal(42, value); // The final values is 42 from the Returns, not 45 from AndDoes
        }
```

Draft documentation:
---

# NS3006

<table>
<tr>
  <td>CheckId</td>
  <td>NS3006</td>
</tr>
<tr>
  <td>Category</td>
  <td>Argument specification</td>
</tr>
</table>

## Cause

Conflicting assignments to out/ref arguments.

## Rule description

A violation of this rule occurs when an out or ref argument is assigned via `CallInfo` in both a `Returns` and an `AndDoes` call.

## How to fix violations

To fix a violation of this rule remove one of the assignments, so the argument is only assigned in either the `Returns` or the `AndDoes` block (not in both).

```csharp
// Incorrect, argument assigned in `Returns` and `AndDoes`:
dictionary.TryGetValue("sample", out Arg.Any<int>())
   .Returns(x => {
       x[1] = 42;
       return true;
    })
   .AndDoes(x => x[1] = 45);

// Correct, only assigned in `AndDoes`:
dictionary.TryGetValue("sample", out Arg.Any<int>())
   .Returns(true)
   .AndDoes(x => x[1] = 45);

// Correct, only assign in `Returns`:
dictionary.TryGetValue("sample", out Arg.Any<int>())
   .Returns(x => {
       x[1] = 42;
       return true;
    });
```

## How to suppress violations

This warning can be suppressed by disabling the warning in the **ruleset** file for the project.
The warning can also be suppressed programmatically for an assembly:
````c#
[assembly: System.Diagnostics.CodeAnalysis.SuppressMessage("Argument specification", "NS3006:Conflicting argument assignments.", Justification = "Reviewed")]
````

Or for a specific code block:
````c#
#pragma warning disable NS3006 // Conflicting argument assignments.
// the code which produces warning
#pragma warning disable NS3006 // Conflicting argument assignments.
